### ISSUE TITLE
Opening the hierarchy for a concept takes a long time

### ISSUE THREAD

**Issue by joelit** [NatLibFi/Skosmos#1377]
## At which URL did you encounter the problem?

https://finto.fi/yso/en/page/p28785

## What steps will reproduce the problem?
1. Load the page with the network inspector open
2. The page content loads in a reasonable time, but the hierarchy takes a long time to open
3. This is shown in the network inspector as the time to finish, for example:
- Chrome: Finish: **5.75 s** DOMContentLoaded: 551 ms Load: 5.67 s
-  Firefox: Finish: **7.06 s** DOMContentLoaded: 583 ms load: 1.01 s
- Edge: Finish: **9.97 s** DOMContentLoaded: 691 ms Load: 9.78 s
(Firefox seems to interpret load time differently to Chrome and Edge)

## What is the expected output? What do you see instead?

This lag time seems to be consistent to certain concept pages. My initial hunch was that this would be proportional to the depth of the concept in the hierarchy tree, but that doesn't seem to be the case. Something like https://finto.fi/yso/en/page/p7873 has more than double the hierarchy depth while keeping the time to finish much shorter.

## What browser did you use? (eg. Firefox, Chrome, Safari, Internet explorer)
Firefox, Chrome & Edge

**Comment by osma** [comment]
How long did the call to the `hierarchy` REST method take? I'm wondering whether most of the time is spent in the REST call to the backend, or in the JS frontend )e.g. jsTree). I assume it's the frontend that is slow, but it wasn't really clear from the report.

**Comment by kinow** [comment]
>How long did the call to the hierarchy REST method take? I'm wondering whether most of the time is spent in the REST call to the backend, or in the JS frontend (e.g. jsTree). I assume it's the frontend that is slow, but it wasn't really clear from the report.

I was curious about that too, so I had a look at the Network tab in Firefox. Three XHR requests finish fairly quickly (~1, 2 seconds):

![image](https://user-images.githubusercontent.com/304786/197976152-dce25f84-7d83-424b-b98f-6b028ef42b20.png)

Looking at Firefox profiler (fancy, it now uses profiler.firefox.com) it shows the code taking a long time to run some natural sorting, which would be a good explanation for the slowness:

![image](https://user-images.githubusercontent.com/304786/197976387-e77093ba-94e5-43c8-8534-c9117efb7ad4.png)

The natural sort function: https://github.com/NatLibFi/Skosmos/blob/1d48d64c738c599b6c7a5c4c44065c84981d204f/resource/js/scripts.js#L338-L354

Maybe someone could try playing with the sorting configuration, or hack the code to remove the natural sort, and see if that would perform better: https://github.com/NatLibFi/Skosmos/blob/e4fd30ea96593658c41ca5a0c900e67b740a366f/resource/js/hierarchy.js#L502-L527

**Comment by mirmaid** [comment]
For users it's a problem that while the hierarchy loads, links and copy-icon do not work, and when the hierarchy has finished loading, the concept view "moves", which is distracting as you have already started to focus on something on it or trying to work with it. Also, if you try to click something several times, you can occasionally break the hierarchy somehow (dotted lines disappear).

**Comment by osma** [comment]
Thanks for the quick analysis @kinow ! Indeed, the natural sort function seems like a probable culprit here.

> the concept view "moves", which is distracting as you have already started to focus on something on it or trying to work with it.

This is partly an unfortunate side effect of the way we reimplemented the custom scrollbar functionality in the latest Skosmos release. When we switched away from the old and unmaintained mCustomScrollbar library, we had to also reimplement the centering of the hierarchy so that the currently selected concept is visible in the hierarchy when you open a concept page. I did that in PR #1372 (and tweaked in #1374) using the standard Element.scrollIntoView method. But this may scroll the entire page in some cases, not just the hierarchy view which was intended. I think we need a proper issue report on this and a separate fix which is unrelated to the performance problem in this issue, although their combined effect (first waiting ~10s for the hierarchy to render, then the auto-scroll happens) is especially bad for usability.

**Comment by osma** [comment]
@kinow I think it's not necessary the `naturalSort` function that is slow here (although it could be), but the related function `nodeLabelSortKey` - this can be seen also in your profiler screenshot, on the last line of the call tree. This one: https://github.com/NatLibFi/Skosmos/blob/1d48d64c738c599b6c7a5c4c44065c84981d204f/resource/js/hierarchy.js#L406-L416

It was introduced in PR #1205 (by yours truly) and intended to extract just the label (not the notation) from the jsTree node representing a concept in the hierarchy. But it performs HTML parsing with jQuery, and I can see now that it could be very slow. Probably we need another, faster mechanism for extracting the label.

**Comment by miguelvaara** [comment]
@kinow @osma, I think the simplest way to reproduce the problem related to the vertical centering of the hierarchy is to come to the page with a direct link, like this: https://finto.fi/ykl/en/page/99.19 (success maybe depends on your screen resolution). By clicking on the hierarchy tree, the problem is hard to reproduce.

**Comment by osma** [comment]
@miguelvaara that problem (separate from this performance issue) is now reported as issue #1382

**Comment by kinow** [comment]
>It was introduced in PR https://github.com/NatLibFi/Skosmos/pull/1205 (by yours truly) and intended to extract just the label (not the notation) from the jsTree node representing a concept in the hierarchy. But it performs HTML parsing with jQuery, and I can see now that it could be very slow. Probably we need another, faster mechanism for extracting the label.

:smile: I suspected that it would be either the natural sort, maybe moving that to the server side. But looks like it's a bit trickier. I hadn't seen that it was parsing HTML. I think you are correct, and on the right path to fix it.

**Comment by osma** [comment]
Yes, it's definitely the HTML parsing. I tried shortcutting it and now the hierarchy loads much faster.
